File: themeablecomponents/src/main/java/com/dertyp7214/themeablecomponents/components/ThemeableEditText.java
package com.dertyp7214.themeablecomponents.components;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.TextView;

import com.dertyp7214.themeablecomponents.utils.OnThemeChangeListener;
import com.dertyp7214.themeablecomponents.utils.Theme;
import com.dertyp7214.themeablecomponents.utils.ThemeManager;

import java.lang.reflect.Field;

import static com.dertyp7214.themeablecomponents.helpers.Utils.getIdFromView;

public class ThemeableEditText extends com.google.android.material.textfield.TextInputEditText {

    private final Context context;
    public OnThemeChangeListener onThemeChangeListener;

    public ThemeableEditText(Context context) {
        super(context);
        this.context = context;
        init();
    }

    public ThemeableEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
        init();
    }

    public ThemeableEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        this.context = context;
        init();
    }

    private void init() {
        ThemeManager themeManager = ThemeManager.getInstance(context);
        onThemeChangeListener = new OnThemeChangeListener() {
            @Override
            public void onThemeChanged(Theme theme, boolean animated) {
                applyTheme(theme);
            }

            @Override
            public ThemeManager.Component.TYPE getType() {
                return ThemeManager.Component.TYPE.EDITTEXT;
            }

            @Override
            public boolean accent() {
                return true;
            }

            @Override
            public String getId() {
                return getIdFromView(ThemeableEditText.this);
            }
        };
        themeManager.register(onThemeChangeListener);
    }

    private void applyTheme(Theme theme) {
        clearCursor();
        setTextHandleColor(theme);
        getBackground().mutate().setColorFilter(theme.getColor(), PorterDuff.Mode.SRC_ATOP);
    }

    private void setTextHandleColor(Theme theme) {
        try {
            Field editorField = TextView.class.getDeclaredField("mEditor");
            if (! editorField.isAccessible()) {
                editorField.setAccessible(true);
            }

            Object editor = editorField.get(this);
            Class<?> editorClass = editor.getClass();

            String[] handleNames =
                    {"mSelectHandleLeft", "mSelectHandleRight", "mSelectHandleCenter"};
            String[] resNames = {"mTextSelectHandleLeftRes", "mTextSelectHandleRightRes",
                    "mTextSelectHandleRes"};

            for (int i = 0; i < handleNames.length; i++) {
                Field handleField = editorClass.getDeclaredField(handleNames[i]);
                if (! handleField.isAccessible()) {
                    handleField.setAccessible(true);
                }

                Drawable handleDrawable = (Drawable) handleField.get(editor);

                if (handleDrawable == null) {
                    Field resField = TextView.class.getDeclaredField(resNames[i]);
                    if (! resField.isAccessible()) {
                        resField.setAccessible(true);
                    }
                    int resId = resField.getInt(this);
                    handleDrawable = getResources().getDrawable(resId);
                }

                if (handleDrawable != null) {
                    Drawable drawable = handleDrawable.mutate();
                    drawable.setColorFilter(theme.getColor(), PorterDuff.Mode.SRC_IN);
                    handleField.set(editor, drawable);
                }
            }
            setHighlightColor(theme.getLightColor());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void clearCursor() {
        try {
            Field f = TextView.class.getDeclaredField("mCursorDrawableRes");
            f.setAccessible(true);
            f.set(this, 0);
        } catch (Exception ignored) {
        }
    }
}

File: themeablecomponents/src/main/java/com/dertyp7214/themeablecomponents/components/ThemeableEditText.kt
@file:Suppress("DEPRECATION")

package com.dertyp7214.themeablecomponents.components

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.widget.TextView
import com.dertyp7214.themeablecomponents.helpers.Utils.getIdFromView
import com.dertyp7214.themeablecomponents.utils.OnThemeChangeListener
import com.dertyp7214.themeablecomponents.utils.Theme
import com.dertyp7214.themeablecomponents.utils.ThemeManager

open class ThemeableEditText : com.google.android.material.textfield.TextInputEditText {

    internal val context: Context
    lateinit var onThemeChangeListener: OnThemeChangeListener

    constructor(context: Context) : super(context) {
        this.context = context
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        this.context = context
        init()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        this.context = context
        init()
    }

    private fun init() {
        val themeManager = ThemeManager.getInstance(context)
        onThemeChangeListener = object : OnThemeChangeListener {
            override val type: ThemeManager.Component.TYPE
                get() = ThemeManager.Component.TYPE.EDITTEXT
            override val id: String
                get() = getIdFromView(this@ThemeableEditText)

            override fun onThemeChanged(theme: Theme, animated: Boolean) {
                applyTheme(theme)
            }

            override fun accent(): Boolean {
                return true
            }
        }
        themeManager.register(onThemeChangeListener)
    }

    private fun applyTheme(theme: Theme) {
        clearCursor()
        setTextHandleColor(theme)
        background.mutate().setColorFilter(theme.color, PorterDuff.Mode.SRC_ATOP)
    }

    private fun setTextHandleColor(theme: Theme) {
        try {
            val editorField = TextView::class.java.getDeclaredField("mEditor")
            if (!editorField.isAccessible) {
                editorField.isAccessible = true
            }

            val editor = editorField.get(this)
            val editorClass = editor.javaClass

            val handleNames = arrayOf("mSelectHandleLeft", "mSelectHandleRight", "mSelectHandleCenter")
            val resNames = arrayOf("mTextSelectHandleLeftRes", "mTextSelectHandleRightRes", "mTextSelectHandleRes")

            for (i in handleNames.indices) {
                val handleField = editorClass.getDeclaredField(handleNames[i])
                if (!handleField.isAccessible) {
                    handleField.isAccessible = true
                }

                var handleDrawable: Drawable? = handleField.get(editor) as Drawable

                if (handleDrawable == null) {
                    val resField = TextView::class.java.getDeclaredField(resNames[i])
                    if (!resField.isAccessible) {
                        resField.isAccessible = true
                    }
                    val resId = resField.getInt(this)
                    handleDrawable = resources.getDrawable(resId)
                }

                if (handleDrawable != null) {
                    val drawable = handleDrawable.mutate()
                    drawable.setColorFilter(theme.color, PorterDuff.Mode.SRC_IN)
                    handleField.set(editor, drawable)
                }
            }
            highlightColor = theme.lightColor
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun clearCursor() {
        try {
            val f = TextView::class.java.getDeclaredField("mCursorDrawableRes")
            f.isAccessible = true
            f.set(this, 0)
        } catch (ignored: Exception) {
        }
    }
}
